site/_posts/2020-11-15-how-to-securely-implement-oauth-rails.md

```diff
@@ -0,0 +1,262 @@
+For this app, we are going to setup three routes:
+* **welcome page**: Our root page or `view`. 
+* **oauth2-callback** endpoint: Receiver of the redirect after logging into FusionAuth.
+* **logout** endpoint: Receiver of the redirect after logging out of FusionAuth.
+
+```ruby
+Rails.application.routes.draw do
+  root to: 'welcome#index'
+  get '/oauth2-callback', to: 'oauth#oauth_callback'
+  get '/logout', to: 'oauth#destroy'
+end
+```
+
+### Authenticating and Logging Out 
+We will start off with a very basic welcome view that contains a link for a user to login to the application.
+
+```ruby
+# app/views/welcome/index.html.erb
+
+<p>Welcome user! Please <%= link_to 'Log In', @login_url %></p>
+```
+
+To begin the login process, we will want to construct a URL to make an [Authorization Code Grant Request](https://fusionauth.io/docs/v1/tech/oauth/endpoints/#authorization-code-grant-request).
+This URL is composed of the `client_id`, `client_secret`, and `redirect_uri` values from our application configuration. 
+
+We will also construct the URL used to make a [Logout](https://fusionauth.io/docs/v1/tech/oauth/endpoints/#logout) request.
+
+```ruby
+# app/controllers/welcome_controller.rb
+
+class WelcomeController < ApplicationController
+  def initialize
+    @login_url = 'http://localhost:9011/oauth2/authorize?client_id=799b3fe1-4a52-494f-a13a-b97b1377e073&response_type=code&redirect_uri=http%3A%2F%2Flocalhost%3A3000%2Foauth2-callback'
+    @logout_url = 'http://localhost:9011/oauth2/logout?client_id=799b3fe1-4a52-494f-a13a-b97b1377e073'
+  end
+end
+```
+
+Now it's time to actually implement the OAuth callback. 
+
+(Note: I utilize `development.rb` for my config properties)
+```ruby
+# app/controllers/oauth_controller.rb
+
+class OauthController < ApplicationController
+  def initialize
+    @client_id = Rails.configuration.x.oauth.client_id
+    @client_secret = Rails.configuration.x.oauth.client_secret
+    @identity_provider_url = Rails.configuration.x.oauth.idp_url
+    @redirect_uri = "http://localhost:3000/oauth2-callback"
+  end
+
+  def oauth_callback
+    code = params[:code]
+    query = { code: code,
+              grant_type: "authorization_code",
+              client_id: @client_id,
+              client_secret: @client_secret,
+              redirect_uri: @redirect_uri }.to_query
+
+    response = RestClient.post("#{@identity_provider_url}oauth2/token?" << query, {})
+    token = JSON.parse(response.body)["access_token"]
+    session[:user_jwt] = {value: token, httponly: true}
+    redirect_to root_path
+  end
+
+  def destroy
+    reset_session
+    redirect_to root_path
+  end
+end
+```
+In summary, `oauth_callback` does a few things:
+1. It handles the redirect following FusionAuth authentication receiving an `authorization code`.
+2. It makes an additional call to exchange the authorization code for an access token.
+3. Saves the `access token` (JWT) on the session to indicate that the user has successfully authenticated.
```

Comment:
Worth noting:

* why did you set the access token cookie to http only (to prevent XSS attacks I'm guessing)
* the discussion in the slack channel notwithstanding, we're using PKCE on all auth code grant examples going forward. Please update to do that.